crates/sui-indexer/src/store/pg_indexer_store.rs

```diff
@@ -492,13 +489,37 @@ impl IndexerStore for PgIndexerStore {
                     .values(events)
                     .execute(conn)?;
 
-                diesel::insert_into(objects::table)
-                    .values(objects)
-                    .execute(conn)?;
-
-                diesel::insert_into(owner_changes::table)
-                    .values(owner_changes)
-                    .execute(conn)?;
+                // Object need to bulk insert by transaction to prevent same object mutated twice in the same sql call,
+                // which will result in "ON CONFLICT DO UPDATE command cannot affect row a second time" error
+                for changes in objects_changes {
+                    diesel::insert_into(objects::table)
+                        .values(&changes.mutated_objects)
+                        .on_conflict(objects::object_id)
+                        .do_update()
+                        .set((
+                            objects::epoch.eq(excluded(objects::epoch)),
+                            objects::checkpoint.eq(excluded(objects::checkpoint)),
+                            objects::version.eq(excluded(objects::version)),
+                            objects::object_digest.eq(excluded(objects::object_digest)),
+                            objects::owner_address.eq(excluded(objects::owner_address)),
+                            objects::previous_transaction.eq(excluded(objects::previous_transaction)),
+                            objects::object_status.eq(excluded(objects::object_status)),
+                        ))
+                        .execute(conn)?;
+
+                    diesel::insert_into(objects::table)
+                        .values(&changes.deleted_objects)
+                        .on_conflict(objects::object_id)
+                        .do_update()
+                        .set((
+                            objects::epoch.eq(excluded(objects::epoch)),
+                            objects::checkpoint.eq(excluded(objects::checkpoint)),
+                            objects::version.eq(excluded(objects::version)),
```

Comment:
this is mainly because we use trigger to populate the owner tables, with insert, we the trigger will receive the old and new data, which we will need to know at which checkpoint the delete was happened, if we just do delete, we won't be able to pass the checkpoint info to the delete statement.

we can delete the "deleted" object part of the archiving job later.